traitsui/qt4/tabular_editor.py

```diff
@@ -913,7 +926,8 @@ def columnResized(self, index, old, new):
             if self._user_widths is None:
                 self._user_widths = [None] * len(self._editor.adapter.columns)
             self._user_widths[index] = new
-            if not self._editor.factory.auto_resize:
+            if (self._editor.factory is not None
```

Comment:
@FedeMiorelli This `if self._editor.factory is not None` check was also made by you in #895. May I ask if this is also needed in the production environment, e.g. running a desktop application manually? Or is it only for tests to pass?  I am trying to understand if there is a more fundamental problem lurking behind. Thank you very much!

Reply:
@kitchoi I had added this check because I'd get a crash when closing the dialog, so while in production desktop application use.

Reply:
@FedeMiorelli Thank you very much. It is good to know this is affecting production use. This definitely worth some investigation. I will look into this.

Reply:
I tried removing this check and then tested this branch against a few simple examples, removing columns and resizing them, and then closing the dialog, but I still could not reproduce the crash.

@FedeMiorelli if you have a simple example that can produce the crash specific to this check, and if you could provide us with the example, that would be very helpful. Without a way to reproduce this in a production environment, I am unable to investigate further nor can I open a separate issue for this. 

The fact that this check is needed in a production environment indicates some deeper issues. However it is not clear to me at this point whether this is an issue in TraitsUI or somewhere else.

Reply:
@kitchoi I have created a minimal case to reproduce the crash when columns are removed, which occurs if the user resizes some column. In this simplified code I have not been able to repeat the other crash however, maybe it was triggered by some other component in the much more complex code this was embedded in.  Sorry if this is not helping!

By the way maybe I could ask a small advice here: you can see I am hacking the TabularAdapter to add a reference to the object (line 15) which i then use to access the column names property. I'm sure there is a smarter / more correct way to do so. I thought I could access the self.object field in the TabularAdapter but it is always None for me.
```

import os, sys
import numpy as np
from traits.api import HasStrictTraits, Int, Property, cached_property, Any
from traitsui.api import View, Group, HGroup, Item, EnumEditor, TextEditor, RangeEditor, TabularEditor, TabularAdapter



class DataTableAdapter ( TabularAdapter ):
    columns      = Property(depends_on='parentObject.DataTableColumns')
    parentObject = Any
    font         = 'Consolas 9'
    alignment    = 'left'
    
    def len ( self, object, trait ):
        self.parentObject = object
        if object is not None and object.DataTable is not None:
            return len(object.DataTable)
        else:
            return 0
    
    @cached_property
    def _get_columns ( self ):
        if self.parentObject is None:
            return []
        else:
            return self.parentObject.DataTableColumns

        
class TestClass(HasStrictTraits):

    NCols = Int(1, label='Columns')
    DataTable = Property(depends_on='NCols')
    DataTableColumns = Property(depends_on='DataTable')

    @cached_property
    def _get_DataTable(self):
        rows = []
        for i in range(4):
            row = np.random.rand(self.NCols)
            rows.append(row)
        
        return rows

    @cached_property
    def _get_DataTableColumns(self):
        if len(self.DataTable) == 0:
            return []
        return [str(i+1) for i in range(len(self.DataTable[0]))]

    traits_view = View(Group(Item('NCols', editor=RangeEditor(low=0, high=10, mode='spinner')),
                                    HGroup(Item('DataTable', editor=TabularEditor(adapter=DataTableAdapter(), drag_move=False, stretch_last_section=False, editable=False, operations=[]),
                                           show_label=False, springy=True), visible_when='len(DataTable)>0'),
                        ),
                         
                   resizable=True,
                   kind='livemodal',
                   width=800, height=300,
                   buttons = [ 'OK', 'Cancel' ],
                  )


if __name__=='__main__':
    TestClass().configure_traits()
```

Reply:
@FedeMiorelli Thank you very much for the example! It is really useful to see how others use the adapter and to know that the attribute error crash is not easily reproducible.

As for the issue with having to create a reference from the adapter back to the object, I wonder if the use of `default_traits_view` is useful here. Instead of defining a static `traits_view`, you can define a view dynamically with a method. Then the `TestClass` can own the adapter and update its columns directly. e.g.:
```
# I added these imports:
from traits.api import Instance, on_trait_change

class DataTableAdapter ( TabularAdapter ):
    font         = 'Consolas 9'
    alignment    = 'left'

    def len ( self, object, trait ):
        if object is not None and object.DataTable is not None:
            return len(object.DataTable)
        else:
            return 0


class TestClass(HasStrictTraits):

    NCols = Int(1, label='Columns')
    DataTable = Property(depends_on='NCols')

    adapter = Instance(DataTableAdapter, kw=dict(columns=["1"]))

    @cached_property
    def _get_DataTable(self):
        rows = []
        for i in range(4):
            row = np.random.rand(self.NCols)
            rows.append(row)

        return rows

    @on_trait_change("NCols")
    def _update_adapter_columns(self):
        self.adapter.columns = [str(i + 1) for i in range(self.NCols)]

    def default_traits_view(self):
        adapter = self.adapter
        return View(Group(Item('NCols', editor=RangeEditor(low=0, high=10, mode='spinner')),
                        HGroup(Item('DataTable', editor=TabularEditor(adapter=adapter, drag_move=False, stretch_last_section=False, editable=False, operations=[]),
                                show_label=False, springy=True), visible_when='len(DataTable)>0'),
            ),
            kind="livemodal",
            resizable=True,
            width=800, height=300,
            buttons = [ 'OK', 'Cancel' ],
        )
```

Reply:
Thanks a lot @kitchoi I learned something new!

Reply:
@FedeMiorelli My pleasure!  I hope this helps and thanks again for looking into this!